File: LeetCode/cpp/2176_Count_Equal_and_Divisible_Pairs_in_an_Array.cpp
class Solution {
public:
	int countPairs(vector<int>& nums, int k) {
		int answer = 0;

		unordered_map<int, vector<int>> um;

		for (int i = 0; i < nums.size(); i++) {
			um[nums[i]].push_back(i);
		}

		for (auto m : um) {
			vector<int> v = m.second;

			for (int i = 0; i < v.size() - 1; i++) {
				for (int j = i + 1; j < v.size(); j++) {
					if ((v[i] * v[j]) % k == 0) {
						answer++;
					}
				}
			}
		}

		return answer;
	}
};

File: LeetCode/java/2176_Count_Equal_and_Divisible_Pairs_in_an_Array.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public int countPairs(int[] nums, int k) {
        int answer = 0;

        Map<Integer, List<Integer>> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (!map.containsKey(nums[i])) {
                map.put(nums[i], new ArrayList<>());
            }

            map.get(nums[i]).add(i);
        }

        for (Map.Entry<Integer, List<Integer>> m : map.entrySet()) {
            List<Integer> list = m.getValue();

            for (int i = 0; i < list.size() - 1; i++) {
                for (int j = i + 1; j < list.size(); j++) {
                    if (list.get(i) * list.get(j) % k == 0) {
                        answer++;
                    }
                }
            }
        }

        return answer;
    }
}